### ISSUE TITLE
Given an  document::element_result, how do I query its type?

### ISSUE THREAD

**Issue by lemire** [simdjson/simdjson#632]
Do I need to try all possible types until I get lucky?

cc @jkeiser

**Comment by jkeiser** [comment]
At the moment, yes :) It's clear we need a type API you can switch on, or a callback API.

**Comment by lemire** [comment]
@jkeiser This is another one where I have to ask: do we want to resolve this for 0.3? It is not part of the milestone.

**Comment by jkeiser** [comment]
This is one we might want to leave for the future. It's not obvious exactly what should be in the enum: number? float / integer? float / integer / unsigned? Do we make it match exactly what we have on tape, possibly restricting the tape implementation in the future?

Further, I don't think it's a huge use case. It's mostly useful when you're implementing a tree-walk like dump(). A tree-walk API, I could see ... "SAX on top of DOM" ... but I think that's also not something we should hold up 0.3 on.

In short, I think we should release what we have, and see exactly how the need for this comes up in the wild. (I think it will, but I think the actual use cases will tell us a lot about how to shape this.)

**Comment by lemire** [comment]
Added to milestone 0.4. We can reassign later.

**Comment by TkTech** [comment]
Hit this updating pysimdjson for 0.3. Elements know their type, why do I have to test until I find it? 

Python can't use the operator overloading to cast, so a bunch of glue functions are needed instead of just returning the most accurate type. Ex:

```
py::class_<dom::element>(m, "element")
        .def_property_readonly("is_null", &dom::element::is_null)
        .def_property_readonly("is_bool",
            [](dom::element &e) {
                auto result = e.get<bool>();
                return !result.error();
            }
        )
...
```

This will impact all bindings for languages that don't have 1:1 types for C++.

**Comment by lemire** [comment]
@jkeiser 
See 
https://github.com/TkTech/pysimdjson/issues/31

This seems quite easy to fix given that the elements already know their type. Can we fix this as part of a 0.3.1 release?

**Comment by jkeiser** [comment]
I hadn't considered the bindings use case when pushing this out, I agree it's important enough to make a release soon (no matter what we call it). 0.3.1 is fine.

Shouldn't be hard. We already even have an internal enum, we could expose it and be done.

**Comment by lemire** [comment]
I have moved it to 0.3.1. 

@jkeiser  Can you handle it? I don't mind doing it, but given that you have been the architect of the new API, you are better qualified.

**Comment by jkeiser** [comment]
Yep, happy to.

**Comment by lemire** [comment]
I have assigned it to you.

**Comment by lemire** [comment]
> This is one we might want to leave for the future. It's not obvious exactly what should be in the enum: number? float / integer? float / integer / unsigned? Do we make it match exactly what we have on tape, possibly restricting the tape implementation in the future?

Well. At a minimum, the JSON specs tells us what we need to support:

- object
- array
- null
- true, false
- numbers
- strings

That might be enough.

It might be possible to do better but the categorization I laid out is straight out of the specification. It is hard to avoid.

Of course, we have several types of numbers... and we may (in time) even have different types of strings (e.g., short strings).

Currently, once you have a number... then it seems we already have enough in the API to figure it out. So I don't think it needs to be hard.

**Comment by jkeiser** [comment]
@lemire I exposed it with concrete types, given that the purpose is helping people write bindings ... I think at least a double vs. int is going to be important for most bindings.

**Comment by jkeiser** [comment]
I can imagine eventually having a generic json_type enum at the top level of simdjson that is just the types in the spec, though.

**Comment by rurban** [comment]
fwiw: jsmn only needs 4 types. OBJECT, HASH, STRING and PRIMITIVE, which contains number, bool and null. Had no problem with that.

**Comment by lemire** [comment]
@rurban Thanks. I think it is easy enough for someone to write a test to see whether an element is either a number a bool or a null using @jkeiser's implementation. You can just do it as a short expression.
